res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1275 @@
+    }
+    // No callback, no light!
+  }
+}
+
+TraktorS4MK2.lightDeck = function(group) {
+  // Freeze the lights while we do this update.
+  this.controller.freeze_lights = true;
+  for (var packet_name in this.controller.OutputPackets) {
+    packet = this.controller.OutputPackets[packet_name];
+    var deck_group_name = "deck1";
+    if (group === "[Channel2]" || group === "[Channel4]") {
+      deck_group_name = "deck2";
+    }
+
+    TraktorS4MK2.lightGroup(packet, deck_group_name, group);
+    TraktorS4MK2.lightGroup(packet, group, group);
+  }
+
+  // FX buttons
+  var packet = this.controller.OutputPackets["output3"];
+  TraktorS4MK2.lightGroup(packet, "[EffectRack1_EffectUnit1]", "[EffectRack1_EffectUnit1]");
+  TraktorS4MK2.lightGroup(packet, "[EffectRack1_EffectUnit2]", "[EffectRack1_EffectUnit2]");
+
+  // Selected deck lights
+  if (group === "[Channel1]") {
+    TraktorS4MK2.controller.setOutput("[Channel1]", "!deck_A", 0x7F, false);
+    TraktorS4MK2.controller.setOutput("[Channel3]", "!deck_C", 0x00, false);
+    TraktorS4MK2.controller.setOutput("deck1", "!deckLight", 0x05, false);
+  } else if (group === "[Channel2]") {
+    TraktorS4MK2.controller.setOutput("[Channel2]", "!deck_B", 0x7F, false);
+    TraktorS4MK2.controller.setOutput("[Channel4]", "!deck_D", 0x00, false);
+    TraktorS4MK2.controller.setOutput("deck2", "!deckLight", 0x05, false);
+  } else if (group === "[Channel3]") {
+    TraktorS4MK2.controller.setOutput("[Channel3]", "!deck_C", 0x7F, false);
+    TraktorS4MK2.controller.setOutput("[Channel1]", "!deck_A", 0x00, false);
+    TraktorS4MK2.controller.setOutput("deck1", "!deckLight", 0x7F, false);
+  } else if (group === "[Channel4]") {
+    TraktorS4MK2.controller.setOutput("[Channel4]", "!deck_D", 0x7F, false);
+    TraktorS4MK2.controller.setOutput("[Channel2]", "!deck_B", 0x00, false);
+    TraktorS4MK2.controller.setOutput("deck2", "!deckLight", 0x7F, false);
+  }
+
+  this.controller.freeze_lights = false;
+  // And now send them all.
+  for (packet_name in this.controller.OutputPackets) {
+    var packet_ob = this.controller.OutputPackets[packet_name];
+    packet_ob.send();
+  }
+}
+
+TraktorS4MK2.pointlessLightShow = function() {
+  var packets = [Object(), Object(), Object()];
+
+  packets[0].length = 53;
+  packets[1].length = 63;
+  packets[2].length = 61;
+
+  for (k = 0; k < 0x7F; k+=0x05) {
+    for (var i = 0; i < packets.length; i++) {
+      packets[i][0] = 0x80 + i;
+      for (j = 1; j < packets[i].length; j++) {
+        packets[i][j] = k;
+      }
+    }
+    controller.send(packets[0], packets[0].length, 0);
+    controller.send(packets[1], packets[1].length, 0);
+    controller.send(packets[2], packets[2].length, 0);
+    // "sleep"
+    var then = Date.now();
+    while (true) {
+      var now = Date.now();
+      if (now - then > 25) {
+        break;
+      }
+    }
+  }
+}
+
+TraktorS4MK2.init = function(id) {
+  TraktorS4MK2.construct()
+  TraktorS4MK2.pointlessLightShow()
```

Comment:
eventually I plan to measure the x/y position of the buttons so I can do more interesting fades / wipes and things :)